AlphaGo/training/supervised_policy_trainer.py

```diff
@@ -6,276 +6,247 @@
-            if self.metadata['generator_sample'] >= self.indices_max:
-                self.shuffle_indices()
+    def next(self):
+        """Iterator function, returning a minibatch of (state, action) pairs.
+        """
+        # Lock must be used for hdf5 access and for computing batch indices.
+        with self.lock:
+            # get indices into data of next minibatch using 'wrapped' indexing.
+            batch_indices = range(self.batch * self.batch_size, (self.batch + 1) * self.batch_size)
+            sample_indices = np.take(self.indices, batch_indices, axis=0, mode='wrap')
+            self.batch += 1
 
-            # return state, action and transformation
-            return state, action, training_sample[1]
+            # Re-shuffle the order of the dataset after each full pass through it.
+            if self.seed is not None and (self.batch % self.batches_per_data_epoch) == 0:
+                self.reshuffle()
 
-    def next(self):
-        state_batch_shape = (self.batch_size,) + self.state_dataset.shape[1:]
-        game_size = state_batch_shape[-1]
-        Xbatch = np.zeros(state_batch_shape)
-        Ybatch = np.zeros((self.batch_size, game_size * game_size))
+            states = self.state_dataset[sample_indices[:, 0], :]
+            actions_xy = self.action_dataset[sample_indices[:, 0], :]
+
+        game_size = states.shape[-1]
+        actions_onehot = np.zeros((self.batch_size, game_size * game_size))
+
+        for i in range(self.batch_size):
+            # Get rotation/reflection symmetry of this sample.
+            transform = ORDERED_TRANSFORMATIONS[sample_indices[i][1]]
+
+            # Transform each plane of the sample. Note that using loop comprehension causes the
+            # transformation to act on the 3rd and 4th dimensions (since they are the non-singleton
+            # dimensions), which are the board rows and columns.
+            states[i] = np.array([transform(plane) for plane in states[i]])
+            actions_onehot[i] = transform(one_hot_action(actions_xy[i], game_size)).flatten()
 
-        for batch_idx in xrange(self.batch_size):
-            state, action, transformation = self.next_indice()
+        return (states, actions_onehot)
 
-            # get rotation symmetry belonging to state
-            transform = BOARD_TRANSFORMATIONS[transformation]
 
-            # get state from dataset and transform it.
-            # loop comprehension is used so that the transformation acts on the
-            # 3rd and 4th dimensions
-            state_transform = np.array([transform(plane) for plane in state])
-            action_transform = transform(one_hot_action(action, game_size))
+class MetadataWriterCallback(Callback):
 
-            Xbatch[batch_idx] = state_transform
-            Ybatch[batch_idx] = action_transform.flatten()
+    def __init__(self, path, initial_data={}):
+        super(MetadataWriterCallback, self).__init__()
+        self.file = path
+        self.metadata = {
```

Comment:
I did struggle with this part. I like to keep my classes as self-contained as possible. Sharing the exact same metadata dict here and externally introduces non-obvious dependencies. Using `update` seemed reasonable since all of the other "metadata" - i.e. the dict created by `validate_args` - is updated once per run, while stuff 'owned' by the callback is updated every epoch.